src/btree/bt_vrfy.c

```diff
@@ -582,6 +583,115 @@ __verify_addr_ts(WT_SESSION_IMPL *session, WT_REF *ref, WT_CELL_UNPACK *unpack,
     return (0);
 }
 
+/*
+ * __verify_btree_id_with_meta --
+ *     With a given btree id, iterate through the metafile table and check if the btree id exists in
+ *     any table. If exists, we place the table URI into a variable Otherwise should return
+ *     WT_NOTFOUND.
+ */
+static int
+__verify_btree_id_with_meta(WT_SESSION_IMPL *session, uint32_t btree_id, const char **uri)
+{
+    WT_CONFIG_ITEM id;
+    WT_CURSOR *meta_cursor;
+    WT_DECL_RET;
+    const char *meta_value;
+
+    WT_ERR(__wt_metadata_cursor(session, &meta_cursor));
+
+    WT_ERR(meta_cursor->reset(meta_cursor));
+    while ((ret = meta_cursor->next(meta_cursor)) == 0) {
+        meta_cursor->get_value(meta_cursor, &meta_value);
+        WT_ERR_NOTFOUND_OK(__wt_config_getones(session, meta_value, "id", &id));
+        if (ret == WT_NOTFOUND) {
+            continue;
+        }
+        if (errno != ERANGE && btree_id == __wt_strtouq(id.str, NULL, 10)) {
+            meta_cursor->get_key(meta_cursor, uri);
+            goto done;
+        }
+    }
+
+done:
+err:
+    WT_TRET(__wt_metadata_cursor_release(session, &meta_cursor));
+    return (ret);
+}
+
+/*
+ * __wt_verify_history_store_tree --
+ *     Verify the history store tree, by iterating and checking all the keys in the history and that
+ *     it exists in both the meta file and the respective data store table. Skips duplicate keys by
+ *     storing the previous key.
+ */
+int
+__wt_verify_history_store_tree(WT_SESSION_IMPL *session)
+{
+    WT_CURSOR *cursor, *data_cursor;
+    WT_DECL_RET;
+    WT_ITEM hs_key;
+    WT_ITEM prev_hs_key;
```

Comment:
I'm not sure how much work is being done in this function, so this may be a worthless idea, but we could avoid doing buffer copies by switching between two `WT_ITEM` keys. So, instead of copying the key to the "prev" `WT_ITEM`, just swap a "current"/"prev" pointer pair that points to the current/prev `WT_ITEM`.

Reply:
I like this approach a lot more, I was actually having trouble with finding the proper way to initialise WT_ITEM (I should have just asked!).

Reply:
I ended up running into issues with this method, however I realised that WT_ITEM's simply contain pointers to both size and data (among some other private details). I managed to attain similar functionality by moving the pointers along at that level. Let me know if the current code doesn't make sense or if it's wasteful memory wise!